### ISSUE TITLE
Underscores in dataset names breaks unique_by

### ISSUE THREAD

**Issue by SirRawlins** [geckoboard/sql-dataset#22]
I've been pushing some data up using this tool today, but kept coming across an error message which said something akin to:

> There was an error sending the data to Geckoboard's API: You cannot use "subscribed_at" in "unique_by" as this field is not defined in the dataset.

This seemed curious to me, as my query, and dataset both have a field named "subscribed_at".

```yml
  - name: subscriptions.all
    update_type: append
    sql: >
      SELECT date_trunc('day', publisher_subscriptions.created_at) subscribed_at,
             publisher_subscriptions.type,
             COUNT(publisher_subscriptions.id)
      FROM publisher_subscriptions
      GROUP BY 1, publisher_subscriptions.type
      ORDER BY 1 DESC
    fields:
      - type: datetime
        name: subscribed_at
      - type: string
        name: type
      - type: number
        name: count
    unique_by:
      - subscribed_at
      - type
```

After downloading and running the debug build, I can see that the compiled request body pulls the underscore out, and uses a key named "subscribedat" - probably during it's serialization through yml, or json.

It seems that using "subscribedat" in my "unique_by" get's things working, but took a while to figure out. 😄

**Comment by jaydiablo** [comment]
I've found that you can force the `key` of the field to what you want it to be, `subscribed_at` in this case, instead of relying on the automatic key generation:

```yaml
- name: subscriptions.all
    update_type: append
    sql: >
      SELECT date_trunc('day', publisher_subscriptions.created_at) subscribed_at,
             publisher_subscriptions.type,
             COUNT(publisher_subscriptions.id)
      FROM publisher_subscriptions
      GROUP BY 1, publisher_subscriptions.type
      ORDER BY 1 DESC
    fields:
      - type: datetime
        name: subscribed_at
        key: subscribed_at
      - type: string
        name: type
      - type: number
        name: count
    unique_by:
      - subscribed_at
```

Then you can use more friendly names for the `name` property, like "Subscribed Day".

**Comment by jnormington** [comment]
Hi @SirRawlins,

Sorry for the troubles you experienced with sql-dataset

We remove characters from the field name to make it a compliant key under the hood.
https://github.com/geckoboard/sql-dataset/blob/master/models/dataset.go#L57-L66

But it appears that same logic isn't applied to unique_by and was an oversight on my part.

As @jaydiablo mentions though you can set a key as part of a field which will instead be used but if the key contains any invalid characters it will also error and be rejected from the server side.

This is a bug in sql-dataset not applying the same logic to unique_by keys as well and I'll try to put a fix in sometime soon - but in the meantime please use the above solution so the fields names display nicely in the Geckoboard dataset UI.

**Comment by jnormington** [comment]
Hi @SirRawlins,

I've made a PR with some fixes and additions for unique by (I fear this also affects optional but will check at sometime). I haven't test manually but from the tests I've written it should work :)

In the PR a builds.zip with version 0.3.0 builds - would be great if you could try one for your OS and feedback. You shouldn't need to do the key thing mentioned above and should work with your "subscribed_at" now but also now you could use a nice name like "Subscribed at" which now should also work correctly for unique by

**Comment by jnormington** [comment]
Hi @SirRawlins,

Thanks for the feedback on the last PR I've created another with new builds - if you get a chance to validate this works as you had originally expected, that would be great.

[builds.zip](https://github.com/geckoboard/sql-dataset/files/2151689/builds.zip)

**Comment by SirRawlins** [comment]
@jnormington thanks so much, that new build works like a charm! 👍

**Comment by jnormington** [comment]
@SirRawlins This is now officially released as version 0.2.3. Which you can get directly from the readme. Thanks again for the bug and help with the resolution.

**Comment by SirRawlins** [comment]
@jnormington you're more than welcome! Thanks for the quick fix. 💪
